core/sr-primitives/src/transaction_validity.rs

```diff
@@ -30,23 +29,240 @@ pub type TransactionLongevity = u64;
+	Invalid(InvalidTransaction),
+	/// Transaction validity can't be determined.
+	Unknown(UnknownTransaction),
+}
+
+impl TransactionValidityError {
+	/// Returns if the reason for the error was block resource exhaustion.
+	pub fn exhausted_resources(&self) -> bool {
+		match self {
+			Self::Invalid(e) => e.exhausted_resources(),
+			Self::Unknown(_) => false,
+		}
+	}
+}
+
+impl Into<&'static str> for TransactionValidityError {
+	fn into(self) -> &'static str {
+		match self {
+			TransactionValidityError::Invalid(invalid) => invalid.into(),
+			TransactionValidityError::Unknown(unknown) => unknown.into(),
+		}
+	}
+}
+
+impl From<InvalidTransaction> for TransactionValidityError {
+	fn from(err: InvalidTransaction) -> Self {
+		TransactionValidityError::Invalid(err)
+	}
+}
+
+impl From<UnknownTransaction> for TransactionValidityError {
+	fn from(err: UnknownTransaction) -> Self {
+		TransactionValidityError::Unknown(err)
+	}
+}
+
+impl Into<crate::ApplyError> for InvalidTransaction {
+	fn into(self) -> crate::ApplyError {
+		TransactionValidityError::from(self).into()
+	}
+}
+
+impl Into<crate::ApplyError> for UnknownTransaction {
+	fn into(self) -> crate::ApplyError {
+		TransactionValidityError::from(self).into()
+	}
+}
+
 /// Information on a transaction's validity and, if valid, on how it relates to other transactions.
 #[derive(Clone, PartialEq, Eq, Encode)]
 #[cfg_attr(feature = "std", derive(Debug))]
 pub enum TransactionValidity {
-	/// Transaction is invalid. Details are described by the error code.
-	Invalid(i8),
+	/// Transaction is invalid.
+	Invalid(InvalidTransaction),
 	/// Transaction is valid.
 	Valid(ValidTransaction),
 	/// Transaction validity can't be determined.
-	Unknown(i8),
+	Unknown(UnknownTransaction),
 }
 
-impl From<Result<ValidTransaction, DispatchError>> for TransactionValidity {
-	fn from(r: Result<ValidTransaction, DispatchError>) -> Self {
-		match r {
-			Ok(v) => TransactionValidity::Valid(v),
-			Err(e) => TransactionValidity::Invalid(e.into()),
+impl Into<TransactionValidity> for TransactionValidityError {
+	fn into(self) -> TransactionValidity {
+		match self {
+			TransactionValidityError::Invalid(invalid) => TransactionValidity::Invalid(invalid),
+			TransactionValidityError::Unknown(unknown) => TransactionValidity::Unknown(unknown),
+		}
+	}
+}
+
+impl Into<TransactionValidity> for InvalidTransaction {
+	fn into(self) -> TransactionValidity {
+		TransactionValidity::Invalid(self)
+	}
+}
+
+impl Into<TransactionValidity> for UnknownTransaction {
+	fn into(self) -> TransactionValidity {
+		TransactionValidity::Unknown(self)
+	}
+}
+
+impl Into<TransactionValidity> for ValidTransaction {
+	fn into(self) -> TransactionValidity {
+		TransactionValidity::Valid(self)
+	}
+}
+
+impl TransactionValidity {
+	/// Combine two `TransactionValidity`s.
+	///
+	/// If both are valid, they are combined.
+	///
+	/// If one of them is not valid, the non-valid one is returned. If both are not valid, `self` is
```

Comment:
Should we prefer `InvalidTransaction` over `UnknownTransaction`? Especially in the context of `exhausted_resources()`. The former indicates the the transaction might still be valid in the future.

Reply:
That said, in the aggregator it probably makes sense to propagate the first error that occurred so maybe it's alright.

Reply:
Before we also always returned the first error. But as we now don't return a Result, I needed to do this manually.